warehouse/oidc/views.py

```diff
@@ -0,0 +1,103 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from pyramid.view import view_config
+
+from warehouse.admin.flags import AdminFlagValue
+from warehouse.macaroons import caveats
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.oidc.interfaces import IOIDCProviderService
+
+
+@view_config(
+    route_name="oidc.mint_token",
+    require_methods=["POST"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=False,
+)
+def mint_token_from_oidc(request):
+    def _invalid(errors):
+        request.response.status = 422
+        return {"message": "Token request failed", "errors": errors}
+
+    oidc_enabled = request.registry.settings[
+        "warehouse.oidc.enabled"
+    ] and not request.flags.enabled(AdminFlagValue.DISALLOW_OIDC)
+    if not oidc_enabled:
+        return _invalid(
+            errors=[
+                {
+                    "code": "not-enabled",
+                    "description": "OIDC functionality not enabled",
+                }
+            ]
+        )
+
+    try:
+        body = request.json_body
+    except ValueError:
+        return _invalid(
+            errors=[{"code": "invalid-json", "description": "missing JSON body"}]
+        )
+
+    unverified_jwt = body.get("token")
+    if not unverified_jwt or not isinstance(unverified_jwt, str):
+        return _invalid(
+            errors=[{"code": "invalid-token", "description": "missing or empty token"}]
+        )
+
+    # For the time being, GitHub is our only OIDC provider.
+    # In the future, this should locate the correct service based on an
+    # identifier in the request body.
+    oidc_service = request.find_service(IOIDCProviderService, name="github")
+    provider = oidc_service.find_provider(unverified_jwt)
+    if not provider:
+        return _invalid(
+            errors=[
+                {"code": "invalid-token", "description": "malformed or invalid token"}
+            ]
+        )
+
+    # At this point, we've verified that the given JWT is valid for the given
+    # project. All we need to do is mint a new token.
+    macaroon_service = request.find_service(IMacaroonService, context=None)
+    not_before = int(time.time())
+    expires_at = not_before + 900
+    serialized, dm = macaroon_service.create_macaroon(
```

Comment:
Flagging for review, since I can't remember if we discussed this: does it make sense to also register a timed task/DB trigger here, to delete this macaroon once it has expired?

I don't think it's a security issue for it to remain (since it'll be invalid), but I can see a couple of advantages to removal:

* We'll be able to emit a `project:api_token:removed` event for these tokens, the way we do for "ordinary" ones
* Without timed removal, these macaroons will be difficult to prune from the DB (since we don't copy the entire caveat set into the DB to filter with after the fact)

Reply:
We probably need to discern between short-lived and long-lived tokens in the events -- if we can do that, I don't think we need to emit a 'removed' event for short-lived tokens.

I think it's probably going to be better to keep these around in order to provide better error messages (and events) in the event that someone tries to use an expired token.

Reply:
Sounds good. I can update the event messages here to emphasize that these are short-lived tokens.

Reply:
(Actually, even better: I can put that in the `additional` metadata, and the event table renderer can look for it and display it as a little "chip" or something like that.)

Reply:
...or just compute that it's "short-lived" from the `expires` key.

Reply:
[b8d6b0c](https://github.com/pypi/warehouse/pull/11272/commits/b8d6b0ca7d0c3875d4a81f1d6a6acfb65bdbeb10) should address this.